### PR TITLE
ENH: stats.unuran: add Polynomial interpolation based numerical inversion

### PR DESCRIPTION
**Pull request by tirthasheshpatel**
#### Reference issue

#14600

#### What does this implement/fix?

Add PINV method from UNU.RAN. Tests, documentation, tutorials, and benchmarks have been added.

#### Additional information

There were some inconsistencies in benchmarks, tests, and typing which have been addressed here.

### CODE REVIEW

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ [hunk 1]
```diff
@@ -918,6 +924,428 @@ cdef class TransformedDensityRejection(Method):
...
+
+    The construction of the interpolation polynomial only works when the PDF is unimodal
+    or when the PDF does not vanish between two modes.
+
+    Parameters
+    ----------
+    dist : object
+        An instance of a class with a ``pdf`` and optionally a ``cdf`` method.
+
+        * ``pdf``: PDF of the distribution. The signature of the PDF is expected to be:
```
Comment (chrisb83) [inline scipy/stats/_unuran/unuran_wrapper.pyx.templ:967]: we need to expain somewhere that there are restrictions on the pdf as in the UNU>RAN documentation>

There are some restrictions for the given distribution:

The support of the distribution (i.e., the region where the PDF is strictly positive) must be connected. In practice this means, that the region where PDF is "not too small" must be connected. Unimodal densities satisfy this condition. If this condition is violated then the domain of the distribution might be truncated.
When the PDF is integrated numerically, then the given PDF must be continuous and should be smooth.
The PDF must be bounded.
The algorithm has problems when the distribution has heavy tails (as then the inverse CDF becomes very steep at 0 or 1) and the requested u-resolution is very small. E.g., the Cauchy distribution is likely to show this problem when the requested u-resolution is less then 1.e-12.

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ [hunk 2]
```diff
@@ -918,6 +924,428 @@ cdef class TransformedDensityRejection(Method):
...
+          signature as the PDF.
+
+    mode : float, optional
+        (Exact) Mode of the distribution. Default is ``None``.
+    center : float, optional
+        Approximate location of the mode or the mean of the distribution. This location
+        provides some information about the main part of the PDF and is used to avoid
+        numerical problems. Default is ``None``.
+    domain : list or tuple of length 2, optional
+        The support of the distribution. Must not contain `nan` values.
```
Comment (chrisb83) [inline scipy/stats/_unuran/unuran_wrapper.pyx.templ:982]: I suggest to remove the comment on nans => we do not mention this for other parameters and it should be clear that nans do not make sense. you could mention that inf is admissible

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ [hunk 3]
```diff
@@ -918,6 +924,428 @@ cdef class TransformedDensityRejection(Method):
...
+        Default is ``None``. When ``None``:
+
+        * If a ``support`` method is provided by the distribution object
+          `dist`, it is used to set the domain of the distribution.
+        * Otherwise the support is assumed to be :math:`(-\infty, \infty)`.
+
+    order : int, optional
+        Order of the interpolating polynomial. Valid orders are between 3 and 17.
+        Higher orders result in fewer intervals for the approximations. Default
+        is 3.
```
Comment (chrisb83) [inline scipy/stats/_unuran/unuran_wrapper.pyx.templ:992]: Default is 5.

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ [hunk 4]
```diff
@@ -918,6 +924,428 @@ cdef class TransformedDensityRejection(Method):
...
+    u_resolution : float, optional
+        Set maximal tolerated u-error. Values of u_resolution must at least 1.e-15 and
+        1.e-5 at most. Notice that the resolution of most uniform random number sources
+        is 2-32= 2.3e-10. Thus a value of 1.e-10 leads to an inversion algorithm that
+        could be called exact. For most simulations slightly bigger values for the
+        maximal error are enough as well. Default is 1e-10.
+    max_intervals : int, optional
+        Set maximum number of intervals. max_ivs must be at least 100 and at most 1000000.
```
Comment (chrisb83) [inline scipy/stats/_unuran/unuran_wrapper.pyx.templ:1000]: maybe: The allowed range is [100, 1_000_000] (to get rid of `max_ivs`)

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ [hunk 5]
```diff
@@ -918,6 +924,428 @@ cdef class TransformedDensityRejection(Method):
...
+        If `random_state` is None (or `np.random`), the `numpy.random.RandomState`
+        singleton is used.
+        If `random_state` is an int, a new ``RandomState`` instance is used,
+        seeded with `random_state`.
+        If `random_state` is already a ``Generator`` or ``RandomState`` instance then
+        that instance is used.
+
+    References
+    ----------
+    .. [1] G. Derflinger, W. Hörmann, and J. Leydold (2008). Numerical inversion when
```
Comment (chrisb83) [inline scipy/stats/_unuran/unuran_wrapper.pyx.templ:1022]: Please use the following citation: ... ACM Transactions on Modeling and Computer Simulation (TOMACS) 20.4 (2010): 1-25.

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ [hunk 6]
```diff
@@ -918,6 +924,428 @@ cdef class TransformedDensityRejection(Method):
...
+    ...        return np.exp(-0.5 * x*x)
+    ... 
+    >>> dist = StandardNormal()
+    >>> urng = np.random.default_rng()
+    >>> rng = NumericalInversePolynomial(dist, random_state=urng)
+
+    Once a generator is created, samples can be drawn from the distribution by calling
+    the `rvs` method:
+
+    >>> rng.rvs()
```
Comment (chrisb83) [inline scipy/stats/_unuran/unuran_wrapper.pyx.templ:1048]: I propose to add a histogram for all the UNURAN methods that we implement => it is the most convincing way to show that rvs works :)

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ [hunk 7]
```diff
@@ -918,6 +924,428 @@ cdef class TransformedDensityRejection(Method):
...
+    >>> approxppf = rng.ppf(u)
+    >>> exactppf = norm.ppf(u)
+    >>> error = np.abs(exactppf - approxppf)
+    >>> plt.plot(u, error)
+    >>> plt.xlabel('u')
+    >>> plt.ylabel('error')
+    >>> plt.title('Error between exact and approximated PPF')
+    >>> plt.show()
+
+    During setup, the CDF of the distribution is computed numerically but it is
```
Comment (chrisb83) [inline scipy/stats/_unuran/unuran_wrapper.pyx.templ:1073]: we could think about moving this to the tutorial, since it is not that important

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ [hunk 8]
```diff
@@ -918,6 +924,428 @@ cdef class TransformedDensityRejection(Method):
...
+    >>> rng = NumericalInversePolynomial(dist, keep_cdf=True, random_state=urng)
+
+    Now, the CDF can be evaluated by calling the `cdf` method:
+
+    >>> rng.cdf(1.959963984540054)
+    0.9750000000042454
+    >>> norm.cdf(1.959963984540054)
+    0.975
+
+    It is also possible to extimate the u-error of the approximated PPF if the exact
```
Comment (chrisb83) [inline scipy/stats/_unuran/unuran_wrapper.pyx.templ:1086]: type: extimate

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ [hunk 9]
```diff
@@ -918,6 +924,428 @@ cdef class TransformedDensityRejection(Method):
...
+    >>> rng = NumericalInversePolynomial(dist, random_state=urng)
+
+    Now, the u-error can be estimated by calling the `u_error` method. It runs a
+    Monte-Carlo simulation to estimate the u-error. By default, 100000 samples are
+    used. To change this, you can pass the number of samples as an argument:
+
+    >>> rng.u_error(sample_size=1000000)  # uses one million samples
+    UError(max_error=8.785994154436594e-11, max_absolute_error=2.930890027826552e-11)
+
+    This returns a namedtuple which contains the maximum u-error and the maximum
```
Comment (chrisb83) [inline scipy/stats/_unuran/unuran_wrapper.pyx.templ:1108]: MAE = mean absolute error. should we say `mae_error` => this is the mean of the absolute errors computed over the given sample

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ [hunk 10]
```diff
@@ -918,6 +924,428 @@ cdef class TransformedDensityRejection(Method):
...
+                    raise UNURANError(self._messages.get())
+        finally:
+            _lock.release()
+            release_unuran_callback(&callback)
+
+    def cdf(self, x):
+        """
+        cdf(x)
+
+        Approximated Cumulative distribution function of the given distribution.
```
Comment (chrisb83) [inline scipy/stats/_unuran/unuran_wrapper.pyx.templ:1236]: cumulative (lower case)

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ [hunk 11]
```diff
@@ -918,6 +924,428 @@ cdef class TransformedDensityRejection(Method):
...
+        ----------
+        sample_size : int, optional
+            Number of samples to use for the estimation. It must be greater than
+            or equal to 1000.
+
+        Returns
+        -------
+        max_error : float
+            Maximum u-error.
+        max_absolute_error : float
```
Comment (chrisb83) [inline scipy/stats/_unuran/unuran_wrapper.pyx.templ:1324]: Mean absolute error

File: scipy/stats/tests/test_sampling.py [hunk 12]
```diff
@@ -5,50 +5,66 @@
...
+    ("TransformedDensityRejection", {"dist": StandardNormal()}),
+    ("DiscreteAliasUrn", {"dist": [0.02, 0.18, 0.8]}),
+    ("NumericalInversePolynomial", {"dist": StandardNormal()})
 ]
 
 
 bad_pdfs_common = [
     # Negative PDF
-    (lambda x: -x, UNURANError, r"50 : PDF\(x\) < 0.!"),
+    (lambda x: -x, UNURANError, r"..."),
```
Comment (chrisb83) [inline scipy/stats/tests/test_sampling.py:61]: why did you change this?

File: scipy/stats/tests/test_sampling.py [hunk 13]
```diff
@@ -582,5 +598,226 @@ def test_bad_urn_factor(self):
...
+    def test_basic(self, dist, mv_ex):
+        rng = NumericalInversePolynomial(dist, random_state=42)
+        check_cont_samples(rng, dist, mv_ex)
+
+    very_slow_dists = ['studentized_range', 'trapezoid', 'triang', 'vonmises',
+                       'levy_stable', 'kappa4', 'ksone', 'kstwo', 'levy_l',
+                       'gausshyper', 'beta', 'anglit']
+    # for some reason, UNU.RAN segmentation faults for the uniform and
+    # arcsine distribution!
+    fatal_fail_dists = ['uniform', 'arcsine']
```
Comment (chrisb83) [inline scipy/stats/tests/test_sampling.py:683]: arcsince is unbounded, we could add it to a list `fails_undbounded_pdf`

not sure why uniform does not work, maybe because it is constant

Comment (chrisb83) [reply scipy/stats/tests/test_sampling.py:683]: `beta` can also be added to unbounded (default b < 1)

**Review by chrisb83 — COMMENTED** [review]
Hi,

I did a first review and it is already in very good shape. I will check the test cases in more details later.

Christoph

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ (hunk 11 above)
Comment (tirthasheshpatel) [reply scipy/stats/_unuran/unuran_wrapper.pyx.templ:1324]: Of course!

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ (hunk 9 above)
Comment (tirthasheshpatel) [reply scipy/stats/_unuran/unuran_wrapper.pyx.templ:1108]: > should we say `mae_error`

As we have been quite explicit in the design of this API, it would be better to cal it `mean_absolute_error` (that's was I was thinking in the first place, `max_absolute_error` is just a silly mistake)

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ (hunk 2 above)
Comment (tirthasheshpatel) [reply scipy/stats/_unuran/unuran_wrapper.pyx.templ:982]: > I suggest to remove the comment on nans => we do not mention this for other parameters and it should be clear that nans do not make sense. you could mention that inf is admissible

You're right. Will make this change.

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ (hunk 1 above)
Comment (tirthasheshpatel) [reply scipy/stats/_unuran/unuran_wrapper.pyx.templ:967]: > we need to expain somewhere that there are restrictions on the pdf as in the UNU>RAN documentation>

I agree. I reserved that for the tutorial. Should we also mention it in the docs?

File: scipy/stats/tests/test_sampling.py (hunk 12 above)
Comment (tirthasheshpatel) [reply scipy/stats/tests/test_sampling.py:61]: > why did you change this?

Different generators throw different errors:

TDR throws `50 : PDF(x) < 0.!` while PINV throws `51 : PDF(center) <= 0.!`

In general, we just want to make sure that an internal error occurs when invalid callbacks are passed to UNU.RAN functions. In that case, it's OK not to validate the message. If you agree, I will mention this point as an inline comment.

Comment (chrisb83) [reply scipy/stats/tests/test_sampling.py:61]: Understood!

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ (hunk 9 above)
Comment (chrisb83) [reply scipy/stats/_unuran/unuran_wrapper.pyx.templ:1108]: Agree, let's go with "mean_absolute_error"

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ (hunk 1 above)
Comment (chrisb83) [reply scipy/stats/_unuran/unuran_wrapper.pyx.templ:967]: I think we should since it is very important to know about the restrictions

Comment (tirthasheshpatel) [reply scipy/stats/_unuran/unuran_wrapper.pyx.templ:983]: Done.

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ (hunk 2 above)
Comment (tirthasheshpatel) [reply scipy/stats/_unuran/unuran_wrapper.pyx.templ]: Done.

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ (hunk 3 above)
Comment (tirthasheshpatel) [reply scipy/stats/_unuran/unuran_wrapper.pyx.templ]: Done.

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ (hunk 4 above)
Comment (tirthasheshpatel) [reply scipy/stats/_unuran/unuran_wrapper.pyx.templ]: Done.

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ (hunk 5 above)
Comment (tirthasheshpatel) [reply scipy/stats/_unuran/unuran_wrapper.pyx.templ]: Done.

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ (hunk 6 above)
Comment (tirthasheshpatel) [reply scipy/stats/_unuran/unuran_wrapper.pyx.templ:1063]: Done.

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ (hunk 7 above)
Comment (tirthasheshpatel) [reply scipy/stats/_unuran/unuran_wrapper.pyx.templ]: Done.

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ (hunk 8 above)
Comment (tirthasheshpatel) [reply scipy/stats/_unuran/unuran_wrapper.pyx.templ]: Done.

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ (hunk 9 above)
Comment (tirthasheshpatel) [reply scipy/stats/_unuran/unuran_wrapper.pyx.templ:1125]: Done.

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ (hunk 10 above)
Comment (tirthasheshpatel) [reply scipy/stats/_unuran/unuran_wrapper.pyx.templ]: Done.

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ (hunk 11 above)
Comment (tirthasheshpatel) [reply scipy/stats/_unuran/unuran_wrapper.pyx.templ]: Done.

File: scipy/stats/tests/test_sampling.py (hunk 13 above)
Comment (tirthasheshpatel) [reply scipy/stats/tests/test_sampling.py]: Done.

**Comment by tirthasheshpatel** [comment]
@chrisb83 I have tried to address your reviews in 6b4b117ba8348f0f45e3a633c15c5aaf0c3acb2a. Also, I have added a tutorial. It's just a first draft.

I still need to document the failures for constant densities. I will add it in the `Notes` section in the documentation and as a note in the tutorials.

File: doc/source/tutorial/stats/sampling_pinv.rst [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+
+Polynomial interpolation based INVersion of CDF (PINV) is an inversion method
+that only requires the density function to sample from a distribution. It is
+based on Polynomial interpolation of the PPF and Gauss-Lobatto integration
+of the PDF. It provides control over the interpolation error and integration
+error. Its primary purpose is to provide very fast sampling which is nearly
+the same for any given distribution at the cost of moderate to slow setup
+time. It is the fastest known inversion method for the fixed-parameter case.
+
+The inversion method is the simplest and most flexible to sample nonuniform
```
Comment (chrisb83) [inline doc/source/tutorial/stats/sampling_pinv.rst:23]: This general introduction could also be moved to the main tutorial on sampling. For example, it is also relevant for HINV.

File: doc/source/tutorial/stats/sampling_pinv.rst [hunk 15]
```diff
@@ -0,0 +1,271 @@
...
+
+This method is suitable for stochastic simulations because of its advantages.
+Some of the most attractive are:
+
+* It preserves the structural properties of the uniform random number sampler.
+* Transforms a uniform random variate :math:`U` one-to-one into non-uniform
+  random variates :math:`X`.
+* Easy and efficient sampling from truncated distributions.
+
+Unfortunately, the PPF is rarely available in closed form or too slow when
```
Comment (chrisb83) [inline doc/source/tutorial/stats/sampling_pinv.rst:40]: If we move the above text to the main tutorial, we could continue the PINV tutorial here

File: doc/source/tutorial/stats/sampling_pinv.rst [hunk 16]
```diff
@@ -0,0 +1,271 @@
...
+available. For many distributions, the CDF is also not easy to obtain. This
+method addresses both the shortcomings. The user only has to provide the PDF
+and optionally a point near the mode (called "center") together with the size
+of the maximal acceptable error. It uses a combination of an adaptive and a
+simple Gauss-Lobatto quadrature to obtain the CDF and Newton's interpolation
+to obtain the PPF. The method is not exact, as it only produces random
+variates of the approximated distribution. Nevertheless, the maximal tolerated
+approximation error can be set close to the machine precision and its relative
```
Comment (chrisb83) [inline doc/source/tutorial/stats/sampling_pinv.rst:48]: what does relative error refer to?

File: doc/source/tutorial/stats/sampling_pinv.rst [hunk 17]
```diff
@@ -0,0 +1,271 @@
...
+error is not much larger than :math:`10^{-15}`. The concept of u-error is used
+to measure and control the error. It is defined as:
```
Comment (chrisb83) [inline doc/source/tutorial/stats/sampling_pinv.rst:50]: TBD if we should also describe the u-error in the general tutorial since it is also relevant for HINV.

File: doc/source/tutorial/stats/sampling_pinv.rst [hunk 18]
```diff
@@ -0,0 +1,271 @@
...
+and :math:`F^{-1}_a` is the approximated PPF of the given distribution.
+
+The maximal u-error is the criterion for approximation errors when calculating
+the CDF and PPF numerically. The maximal tolerated u-error of an algorithm is
+called the u-resolution of the algorithm and denoted by :math:`\epsilon_{u}`:
+
+.. math::
+    \sup_{u \in (0,1)} | u - F\left(F^{-1}_{a}(u)\right) | \le \epsilon_{u}
+
+There are two drawbacks of using u-resolution:
```
Comment (chrisb83) [inline doc/source/tutorial/stats/sampling_pinv.rst:65]: That sounds a bit negative :) I would say the following:

The main advantage of the u-error is that it can be easily computed if the CDF is available. We refer to [cite PINV paper] for a more detailed discussion.

Maybe we should even remove the drawbacks: it is probably not easy to understand why high/narrow peaks are a problem etc and we already explain that the method does not work well in that case

File: doc/source/tutorial/stats/sampling_pinv.rst [hunk 19]
```diff
@@ -0,0 +1,271 @@
...
+    >>> from scipy.stats import norm
+    >>> from scipy.stats import NumericalInversePolynomial
+    >>> class StandardNormal:
+    ...     def pdf(self, x):
+    ...         return np.exp(-0.5 * x*x)
+    ...
+    >>> dist = StandardNormal()
+    >>> urng = np.random.default_rng()
+    >>> rng = NumericalInversePolynomial(dist, random_state=urng)
+    >>> rvs = rng.rvs(100000)
```
Comment (chrisb83) [inline doc/source/tutorial/stats/sampling_pinv.rst:121]: maybe that number can be reduced quite a bit since we only use 50 bins, e.g. I guess the histogram still looks ok with ize=10,000?

File: doc/source/tutorial/stats/sampling_pinv.rst [hunk 20]
```diff
@@ -0,0 +1,271 @@
...
+of subintervals required to achieve the error goal which saves memory and
+makes sampling fast. `NumericalInverseHermite` is a similar inversion method
+that inverts the CDF based on Hermite interpolation and provides control
+over the maximum tolerated error via u-resolution. But it makes use of a lot
+more intervals compared to `NumericalInversePolynomial`:
+
+    >>> from scipy.stats import NumericalInverseHermite
+    >>> # NumericalInverseHermite accepts a `tol` parameter to set the
+    >>> # u-resolution of the generator.
+    >>> rng_hermite = NumericalInverseHermite(norm(), tol=1e-12)
```
Comment (chrisb83) [inline doc/source/tutorial/stats/sampling_pinv.rst:191]: i just noted the slight inconsistency in the interface:

tol vs u_resolution
intervals vs n_intervals

just mentioning it here to start a discussion

File: doc/source/tutorial/stats/sampling_pinv.rst [hunk 21]
```diff
@@ -0,0 +1,271 @@
...
+    >>> x = np.linspace(-10, 10, num=100_000)
+    >>> x_error = np.abs(dist.cdf(x) - rng.cdf(x))
+    >>> x_error.max()
+    4.506062190046123e-12
+    >>> I0 = np.sqrt(2*np.pi)
+    >>> max_integration_error = 0.05 * I0 * 1e-10
+    >>> x_error.max() <= max_integration_error
+    True
+
+Note that each CDF evaluation uses a Gauss-Lobatto quadrature which calls
```
Comment (chrisb83) [inline doc/source/tutorial/stats/sampling_pinv.rst:270]: I would still mention that the CDF values computed in the setup are used and that only some further fine-tuning is needed, i.e. no need to do a full integration over the PDF

**Review by chrisb83 — COMMENTED** [review]
I reviewed the tutorial

File: doc/source/tutorial/stats/sampling_pinv.rst (hunk 14 above)
Comment (tirthasheshpatel) [reply doc/source/tutorial/stats/sampling_pinv.rst:23]: I was thinking of expanding the main tutorials to explain each of the methods separately instead of explaining them in a single paragraph. Let me know what you think about that. I can open a PR in parallel where we can track those changes.

File: doc/source/tutorial/stats/sampling_pinv.rst (hunk 16 above)
Comment (tirthasheshpatel) [reply doc/source/tutorial/stats/sampling_pinv.rst:48]: I think the relative x-error:

image

File: doc/source/tutorial/stats/sampling_pinv.rst (hunk 18 above)
Comment (tirthasheshpatel) [reply doc/source/tutorial/stats/sampling_pinv.rst:65]: > Maybe we should even remove the drawbacks: it is probably not easy to understand why high/narrow peaks are a problem etc and we already explain that the method does not work well in that case

Yup, this sounds good. I will add the advantage instead.

File: doc/source/tutorial/stats/sampling_pinv.rst (hunk 19 above)
Comment (tirthasheshpatel) [reply doc/source/tutorial/stats/sampling_pinv.rst:121]: > I guess the histogram still looks ok with ize=10,000?

Yes, it does. Will change it this way.

File: doc/source/tutorial/stats/sampling_pinv.rst (hunk 20 above)
Comment (tirthasheshpatel) [reply doc/source/tutorial/stats/sampling_pinv.rst:191]: > tol vs u_resolution
> intervals vs n_intervals

I also wanted to discuss this: `u_resolution` is much more descriptive than `tol` and it aligns with the naming convention of the paper and UNU.RAN. So, should we deprecate `tol` and use `u_resolution` instead? What do you think, @mdhaber?

I will change `n_intervals` to just `intervals` as both are descriptive enough.

File: doc/source/tutorial/stats/sampling_pinv.rst (hunk 21 above)
Comment (tirthasheshpatel) [reply doc/source/tutorial/stats/sampling_pinv.rst:270]: > I would still mention that the CDF values computed in the setup are used and that only some further fine-tuning is needed, i.e. no need to do a full integration over the PDF

Sure, will mention this!

File: doc/source/tutorial/stats/sampling_pinv.rst (hunk 18 above)
Comment (tirthasheshpatel) [reply doc/source/tutorial/stats/sampling_pinv.rst]: Done.

File: doc/source/tutorial/stats/sampling_pinv.rst (hunk 19 above)
Comment (tirthasheshpatel) [reply doc/source/tutorial/stats/sampling_pinv.rst]: Done.

File: doc/source/tutorial/stats/sampling_pinv.rst (hunk 20 above)
Comment (tirthasheshpatel) [reply doc/source/tutorial/stats/sampling_pinv.rst:211]: Changed `n_intervals` to just `intervals`.

File: doc/source/tutorial/stats/sampling_pinv.rst (hunk 21 above)
Comment (tirthasheshpatel) [reply doc/source/tutorial/stats/sampling_pinv.rst]: Done.

File: doc/source/tutorial/stats/sampling_pinv.rst (hunk 14 above)
Comment (chrisb83) [reply doc/source/tutorial/stats/sampling_pinv.rst:23]: Agree, we can keep it as it is and expand the main tutorial as a follow-up

File: doc/source/tutorial/stats/sampling_pinv.rst (hunk 15 above)
Comment (chrisb83) [reply doc/source/tutorial/stats/sampling_pinv.rst:40]: (follow-up)

File: doc/source/tutorial/stats/sampling_pinv.rst (hunk 16 above)
Comment (chrisb83) [reply doc/source/tutorial/stats/sampling_pinv.rst:48]: I am still confused about this: we only control the u-error, not the x-error. Is the rel x-error really below 1e-15?

File: doc/source/tutorial/stats/sampling_pinv.rst [hunk 22]
```diff
@@ -0,0 +1,300 @@
...
+* When the PDF is integrated numerically, then the given PDF must be
+  continuous and should be smooth.
+* The PDF must be bounded.
+* The algorithm has problems when the distribution has heavy tails (as then
+  the inverse CDF becomes very steep at 0 or 1) and the requested u-resolution
+  is very small. E.g., the Cauchy distribution is likely to show this problem
+  when the requested u-resolution is less then 1.e-12.
+
+.. warning::
+    This method does not work for densities with constant parts (e.g.
```
Comment (chrisb83) [inline doc/source/tutorial/stats/sampling_pinv.rst:88]: I think the unbounded arcsine density also leads to a seg fault, right? Should we mention it?

File: doc/source/tutorial/stats/sampling_pinv.rst [hunk 23]
```diff
@@ -0,0 +1,300 @@
...
+  this step is skipped. Otherwise, the ends of the tails are cut off such that
+  the area under them is less than or equal to :math:`0.05\epsilon_{u}`.
+* The domain is divided into subintervals to compute the CDF and PPF.
+* The CDF is computed using Gauss-Lobatto quadrature such that the integration
+  error is at most :math:`0.05I_{0}\epsilon_{u}` where :math:`I_{0}` is
+  approximately the total area under the PDF.
+* The PPF is computed using Newton's interpolating formula with maximum
+  interpolation error :math:`0.9\epsilon_{u}`.
+
+To initialize the generator to sample from a stnadard normal distribution, do:
```
Comment (chrisb83) [inline doc/source/tutorial/stats/sampling_pinv.rst:106]: Typo: stndard

File: doc/source/tutorial/stats/sampling_pinv.rst (hunk 16 above)
Comment (tirthasheshpatel) [reply doc/source/tutorial/stats/sampling_pinv.rst:48]: > I am still confused about this: we only control the u-error, not the x-error. Is the rel x-error really below 1e-15?

The paper mentions it. Though, I didn't see any proof. This might be a heuristic based on their experiments.

File: doc/source/tutorial/stats/sampling_pinv.rst (hunk 20 above)
Comment (chrisb83) [reply doc/source/tutorial/stats/sampling_pinv.rst:211]: I am also in favour of u_resolution

File: doc/source/tutorial/stats/sampling_pinv.rst (hunk 22 above)
Comment (tirthasheshpatel) [reply doc/source/tutorial/stats/sampling_pinv.rst:88]: > I think the unbounded arcsine density also leads to a seg fault, right?

Tested it locally again and it doesn't fail now. Might be a flaky failure while testing locally. I have removed it failing distributions. Let's see if the CI passes.

File: doc/source/tutorial/stats/sampling_pinv.rst (hunk 23 above)
Comment (tirthasheshpatel) [reply doc/source/tutorial/stats/sampling_pinv.rst]: Done.

File: scipy/stats/tests/test_sampling.py [hunk 24]
```diff
@@ -582,5 +597,230 @@ def test_bad_urn_factor(self):
...
+         [-np.inf, np.nan, np.inf], 0,
+         [[np.nan, 0.5, 0.1], [0.2, 0.4, np.inf], [-2, 3, 4]]]
+
+    @pytest.mark.parametrize("u", u)
+    def test_ppf(self, u):
+        dist = StandardNormal()
+        rng = NumericalInversePolynomial(dist, u_resolution=1e-14)
+        # Older versions of NumPy throw RuntimeWarnings for comparisons
+        # with nan.
+        with suppress_warnings() as sup:
```
Comment (chrisb83) [inline scipy/stats/tests/test_sampling.py:735]: this might be a bit confusing. I guess the idea is to check the following scenarios:

- for finite values, use all_close to compare the finite values
- for special cases like values outside of [0,1], I guess the result should be nan (?)
- ppf(0) = -inf, ppf(1) = inf
- nans and empty values map to nan (?)

maybe add a brief comment inline or split up the test?

**Review by chrisb83 — COMMENTED** [review]
There is not much work left in my view 👍 I just had one remark on the documentation.
The test cases look very comprehensive, I just mentioned a few minor observations

File: scipy/stats/tests/test_sampling.py (hunk 24 above)
Comment (tirthasheshpatel) [reply scipy/stats/tests/test_sampling.py:735]: > maybe add a brief comment inline or split up the test?

Adding an inline comment for the purpose of each test case sounds good. I will do that.

File: doc/source/tutorial/stats/sampling_pinv.rst (hunk 16 above)
Comment (chrisb83) [reply doc/source/tutorial/stats/sampling_pinv.rst:48]: Hm, I did not see it right away in the paper. Could you point me to it?

Comment (tirthasheshpatel) [reply doc/source/tutorial/stats/sampling_pinv.rst:48]: I think the two highlighted sentences say that the relative x-error achieved by the algorithm is near to machine precision (1e-15):

image

Though, the second highlighted sentence might be talking about the algorithm that uses the x-error as the criterion. What do you think?

File: scipy/stats/tests/test_sampling.py (hunk 24 above)
Comment (tirthasheshpatel) [reply scipy/stats/tests/test_sampling.py:746]: Done.

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ [hunk 25]
```diff
@@ -918,6 +924,452 @@ cdef class TransformedDensityRejection(Method):
...
+
+    >>> import matplotlib.pyplot as plt
+    >>> u = np.linspace(0.01, 0.99, 1000)
+    >>> approxppf = rng.ppf(u)
+    >>> exactppf = norm.ppf(u)
+    >>> error = np.abs(exactppf - approxppf)
+    >>> plt.plot(u, error)
+    >>> plt.xlabel('u')
+    >>> plt.ylabel('error')
+    >>> plt.title('Error between exact and approximated PPF')
```
[the remark this answers is not in the document]
Comment (tirthasheshpatel) [reply scipy/stats/_unuran/unuran_wrapper.pyx.templ]: Done.

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ [hunk 26]
```diff
@@ -918,6 +924,452 @@ cdef class TransformedDensityRejection(Method):
...
+
+    As this is an invertion method, the approximated PPF can be evaluated by calling
+    the `ppf` method:
+
+    >>> rng.ppf(0.975)
+    1.9599639857012559
+    >>> norm.ppf(0.975)
+    1.959963984540054
+
+    We can see how well the generator approximates the standard normal distribution by
```
[the remark this answers is not in the document]
Comment (tirthasheshpatel) [reply scipy/stats/_unuran/unuran_wrapper.pyx.templ]: Done.

File: doc/source/tutorial/stats/sampling_pinv.rst (hunk 22 above)
Comment (chrisb83) [reply doc/source/tutorial/stats/sampling_pinv.rst:88]: Interesting, `arcsine` is just `beta(0.5,0.5)`. Maybe the tails at 0 and 1 are cut off in a way that PINV works in that case.

File: doc/source/tutorial/stats/sampling_pinv.rst (hunk 16 above)
Comment (chrisb83) [reply doc/source/tutorial/stats/sampling_pinv.rst:48]: i think the authors mean the following:

- if the u-error is ~1e-15, they will call the algorithm exact
- then they explain that the absolute x-error has some limitations (approximation needs to be very accurate in the tails)
- if the relative x-error is used, you avoid the problem in the fails but you get a problem at 0

They therefore conclude that the u-error is the right measure (only the absolute one is considered since u is in (0,1) in the expression |u -F(...)|)

and the u-error and x-error are related:
image

but in general, you cannot guarantee that the x-error is small if the u-error is

if we just say "Nevertheless, the maximal tolerated approximation error can be set close to the machine precision", it should be fine.

Comment (tirthasheshpatel) [reply doc/source/tutorial/stats/sampling_pinv.rst:48]: I think I understand it now. Thanks for clearing that out!

> if we just say "Nevertheless, the maximal tolerated approximation error can be set close to the machine precision", it should be fine.

Agreed. I will change it this way.

Comment (tirthasheshpatel) [reply doc/source/tutorial/stats/sampling_pinv.rst]: Changed now. Thanks, @chrisb83!

**Review by chrisb83 — APPROVED** [review]
My comments have been addressed. I checked the documentation / tutorial and did not find any issues.

File: doc/source/tutorial/stats/sampling_pinv.rst (hunk 20 above)
Comment (tirthasheshpatel) [reply doc/source/tutorial/stats/sampling_pinv.rst:210]: Let me know if you agree, @mdhaber. If you do, I think this is good to go and we can start working on shifting `NumericalInverseHermite` to UNU.RAN.

Comment (mdhaber) [reply doc/source/tutorial/stats/sampling_pinv.rst:210]: I'm not really convinced, but I'm not blocking anything. I'm that case, when you work on `NumericalInverseHermite`, consider deprecating `tol` and accepting `u_resolution`, then.

Comment (chrisb83) [reply doc/source/tutorial/stats/sampling_pinv.rst:210]: ok, looks like we can proceed with `u_resolution` for PINV. I would also prefer to change that in `NumericalInverseHermite` (`tol` is not a bad option either). Compared to  `random_state` discussion, this is a very small change that is easy to implement.

**Comment by chrisb83** [comment]
Thanks for adding this method, Tirth!
